File: src/homebrew.rs
use anyhow::{Context, Result};
use std::path::PathBuf;
use std::process::Command;

/// Check if Homebrew is installed
pub fn is_installed() -> bool {
    Command::new("brew")
        .arg("--version")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

/// Check if a package is installed
pub fn is_package_installed(package: &str) -> Result<bool> {
    let output = Command::new("brew").args(["list", package]).output()?;

    Ok(output.status.success())
}

/// Install a package (with optional unlinking for isolation)
pub fn install_package(package: &str, link: bool) -> Result<()> {
    use colored::*;

    println!("  Installing {} via Homebrew...", package.cyan());

    let status = Command::new("brew")
        .args(["install", package])
        .status()
        .context("Failed to run 'brew install'")?;

    if !status.success() {
        anyhow::bail!("Failed to install {}", package);
    }

    // If we don't want it linked (pure package), unlink it
    if !link {
        let _ = unlink_package(package); // Unlink after install
    }

    Ok(())
}

/// Unlink a package (remove from global availability)
pub fn unlink_package(package: &str) -> Result<()> {
    let output = Command::new("brew")
        .args(["unlink", package])
        .output()
        .context("Failed to run 'brew unlink'")?;

    if !output.status.success() {
        anyhow::bail!("Failed to unlink {}", package);
    }

    // Only show output if symlinks were actually removed (not "0 symlinks removed")
    let stdout = String::from_utf8_lossy(&output.stdout);
    if !stdout.contains("0 symlinks removed") {
        print!("{}", stdout);
    }

    Ok(())
}

/// Get the installation path for a package
pub fn package_prefix(package: &str) -> Result<PathBuf> {
    let output = Command::new("brew")
        .args(["--prefix", package])
        .output()
        .context(format!("Failed to get prefix for {}", package))?;

    if !output.status.success() {
        anyhow::bail!("Package {} is not installed", package);
    }

    let prefix = String::from_utf8(output.stdout)?.trim().to_string();

    Ok(PathBuf::from(prefix))
}

/// Uninstall a package
pub fn uninstall_package(package: &str) -> Result<()> {
    let output = Command::new("brew")
        .args(["uninstall", package])
        .output()
        .context("Failed to run 'brew uninstall'")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("{}", stderr.trim());
    }

    Ok(())
}

/// Run brew cleanup
pub fn cleanup() -> Result<()> {
    let status = Command::new("brew")
        .arg("cleanup")
        .status()
        .context("Failed to run 'brew cleanup'")?;

    if !status.success() {
        anyhow::bail!("brew cleanup failed");
    }

    Ok(())
}

/// Ensure a package is installed (install if needed)
/// Returns the package prefix path
pub fn ensure_package(package: &str, link: bool) -> Result<PathBuf> {
    if !is_package_installed(package)? {
        install_package(package, link)?;
    } else if !link {
        // If package exists but we want it unlinked, unlink it
        let _ = unlink_package(package); // Ignore error if already unlinked
    }

    package_prefix(package)
}

/// Get list of dependencies for a package
pub fn package_deps(package: &str) -> Result<Vec<String>> {
    let output = Command::new("brew")
        .args(["deps", "--formula", package])
        .output()
        .context(format!("Failed to get dependencies for {}", package))?;

    if !output.status.success() {
        return Ok(Vec::new()); // No dependencies or package doesn't exist
    }

    let deps_output = String::from_utf8(output.stdout)?;
    let deps: Vec<String> = deps_output
        .lines()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    Ok(deps)
}

/// Check if a tap is already tapped
pub fn is_tap_tapped(tap: &str) -> Result<bool> {
    let output = Command::new("brew")
        .args(["tap"])
        .output()
        .context("Failed to run 'brew tap'")?;

    if !output.status.success() {
        return Ok(false);
    }

    let taps_output = String::from_utf8(output.stdout)?;
    Ok(taps_output.lines().any(|line| line.trim() == tap))
}

/// Add a tap
pub fn tap(tap_name: &str) -> Result<()> {
    use colored::*;

    println!("  Tapping {} via Homebrew...", tap_name.cyan());

    let status = Command::new("brew")
        .args(["tap", tap_name])
        .status()
        .context("Failed to run 'brew tap'")?;

    if !status.success() {
        anyhow::bail!("Failed to tap {}", tap_name);
    }

    Ok(())
}

/// Remove a tap
pub fn untap(tap_name: &str) -> Result<()> {
    use colored::*;

    println!("  Untapping {} via Homebrew...", tap_name.cyan());

    let status = Command::new("brew")
        .args(["untap", tap_name])
        .status()
        .context("Failed to run 'brew untap'")?;

    if !status.success() {
        anyhow::bail!("Failed to untap {}", tap_name);
    }

    Ok(())
}

File: lib/macdev/homebrew.rb
# frozen_string_literal: true

require 'English'

module Macdev
  # Homebrew package manager integration
  class Homebrew
    def self.installed?
      system('which brew > /dev/null 2>&1')
    end

    def self.install(package, link: true)
      # Install the package normally
      system('brew', 'install', package) || abort("\e[31mFailed to install #{package}\e[0m")

      # If we don't want it linked (pure package), unlink it after install
      unlink(package) unless link

      # Return the install path
      prefix(package)
    end

    def self.install_cask(package)
      system('brew', 'install', '--cask', package) ||
        abort("\e[31mFailed to install cask #{package}\e[0m")
    end

    def self.prefix(package)
      output = `brew --prefix #{package} 2>/dev/null`.strip
      return nil if output.empty? || $CHILD_STATUS.exitstatus != 0

      output
    end

    def self.unlink(package)
      system('brew', 'unlink', package, out: File::NULL, err: File::NULL)
    end

    def self.package_installed?(package)
      system("brew list #{package} > /dev/null 2>&1")
    end

    def self.uninstall_package(package)
      output = `brew uninstall #{package} 2>&1`
      success = $CHILD_STATUS.exitstatus.zero?

      puts "    \e[33m⚠\e[0m Failed to uninstall: #{output.strip}" unless success

      success
    end

    def self.uninstall_cask(cask)
      output = `brew uninstall --cask #{cask} 2>&1`
      success = $CHILD_STATUS.exitstatus.zero?

      puts "    \e[33m⚠\e[0m Failed to uninstall: #{output.strip}" unless success

      success
    end

    def self.cask_installed?(cask)
      system("brew list --cask #{cask} > /dev/null 2>&1")
    end

    def self.cleanup
      system('brew', 'cleanup')
    end

    def self.tap_tapped?(tap)
      system("brew tap | grep -q '^#{tap}$'")
    end

    def self.tap(tap)
      system('brew', 'tap', tap) || abort("\e[31mFailed to tap #{tap}\e[0m")
    end

    def self.untap(tap)
      system('brew', 'untap', tap) || abort("\e[31mFailed to untap #{tap}\e[0m")
    end

    def self.package_info(package)
      require 'json'

      output = `brew info --json=v2 #{package} 2>&1`
      return nil if $CHILD_STATUS.exitstatus != 0

      begin
        data = JSON.parse(output)
        formulae = data['formulae']
        return nil if formulae.nil? || formulae.empty?

        formula = formulae[0]
        {
          version: formula['versions']['stable'],
          formula: formula['full_name']
        }
      rescue JSON::ParserError, NoMethodError
        nil
      end
    end

    def self.package_deps(package)
      output = `brew deps --formula #{package} 2>&1`
      return [] if $CHILD_STATUS.exitstatus != 0

      output.lines.map(&:strip).reject(&:empty?)
    end
  end
end
